### ISSUE TITLE
Pupul music glitches every second

### ISSUE THREAD

**Issue by vinriviere** [aranym/aranym#68]
Heard there:
https://www.youtube.com/watch?v=GOkDuLmgWFo&t=2984

We can hear a small audio glitch every second in the Pupul music. Sorry for the bad bug report, the issue is not much audible in the above video, as I speak over the music. Turn volume up, or try yourself, that's easy.

I didn't make further tests, so I don't currently know if the problem comes from Pupul, FreeMiNT, EmuTOS or ARAnyM, of a combination of them. Nevertheless, I open this issue to track investigation progress.

**Comment by vinriviere** [comment]
Daroou (Pupul's author) gave me further information. The replay routine comes from @huth, using Falcon audio functions. It juggles between 2 DMA buffers. buffptr() is used to determine the current playing position. Daroou says that buffptr() is not correctly implemented with ARAnyM (or EmuTOS?).

One difference from real Falcon hardware is here: Call buffptr() to determine approximately when you are in the middle of the first buffer. Switch to the second buffer with setbuffer(), then immediately call buffptr(). On real Falcon hardware, buffptr() only provides a pointer to somewhere inside the second buffer after the fist buffer has been completely played. While on ARAnyM/EmuTOS, setbuffer() seems to operate immediately.

Maybe @mikrosk has an opinion about this issue.

**Comment by th-otto** [comment]
ARAnyM does not attempt to emulate buffptr(), if at all it tries to emulate the hardware. 

Does Pupul use the normal XBIOS functions, so that ZMagxSnd, and the aranym driver is used?

**Comment by vinriviere** [comment]
See above. I uses normal XBIOS functions. As you can see in the video, the setup is just official ARAnyM (release) + latest official FreeMiNT snapshot, without further configuration. As far as I know, ZMagxSnd is not used on such setup.

Maybe EmuTOS does some weird sound implementation on ARAnyM? Sorry, I'm not familiar with that code.

**Comment by huth** [comment]
One easy thing that you could check: Does  the problem also occur with the original Falcon TOS 4.04 ? If so, then this is likely a problem in Aranym's DMA sound emulation. If not, then there is likely something wrong in the XBIOS sound functions from EmuTOS.

**Comment by th-otto** [comment]
There are a few differences in how buffptr and setbuffer implemented between EmuTOS/TOS (dunno if anything of these causes any problems)

- In TOS, buffptr reads the play register first, then the record register. That has the effect, that at the end of the routine, the record registers are selected. EmuTOS reads the record registers first, so at the end the play registers are selected

- in TOS, access to the frame counter register is done using word move, from the even address. EmuTOS uses bytes moves from the odd address.

- in both functions, TOS reads/writes the least significant byte first (L/M/H), EmuTOS reads/writes the most significant value first (H/M/L)

**Comment by mikrosk** [comment]
OK, I digged in my mailbox. When I had been coding [mxplay](https://sourceforge.net/projects/mxplay/files/latest/download) and testing it in Aranym, the story went like this:

- there wasn't any support for XBIOS so I had to use `zmagxsnd.prg` what was a real pain because it didn't work with memory protection
- with this "driver" I had to make a special hack: `Buffptr()` didn't work very reliably (it was slightly sliding off) so I had to use Timer A to get the idea when to swap buffer
- in the beginning of 2013 I convinced @huth to add those XBIOS functions directly to EmuTOS so `zmagxsnd.prg` wasn't needed
- in 09/2017 I tested it with mxPlay again and to my shock, Aranym got much better in this area -- IIRC it was @pmandin who made the `Buffptr()` fixes -- I didn't need my Timer A hacks anymore!
- then, at some point, it got broken again but broken in a horrible way -- neither Timer A nor `Buffptr` method works anymore!

> ARAnyM does not attempt to emulate buffptr(), if at all it tries to emulate the hardware.

It does and it does. Well, it used to at least. So it's a regression.

**Comment by mikrosk** [comment]
I'm happy to report that latest Aranym as of today works absolutely wonderfully. SDL1 + latest master + either Timer-A/Buffptr() based replay in mxPlay works a treat!

P.S. I also tested Pupul, works equally great. Thank you so much for fixing this!

**Comment by vinriviere** [comment]
I've just tested with the latest ARAnyM snapshot. Same issue 🙁

I got it there:
https://bintray.com/aranym/aranym-files/snapshots#files/snapshots
For memories the filename was:
aranym-1.1.0-cygwin-i386-4ebb1861-sdl1.zip
Date of executables is 02/07/2020.

**Comment by th-otto** [comment]
Looks like that was an older release. Latest one is https://bintray.com/aranym/aranym-files/download_file?file_path=snapshots%2Faranym-1.1.0-cygwin-i386-76892d52-sdl1.zip dated 24/07/2020

**Comment by vinriviere** [comment]
I urge the ARAnyM team to put a link to latest snapshots somewhere. This is a real pain.

**Comment by vinriviere** [comment]
https://bintray.com/aranym/aranym-files/snapshots#files/snapshots
I actually clicked on the bottom one on the right... while the latest is actually the top one! That Bintray UI without any timestamp is utterly bad.

**Comment by th-otto** [comment]
Yes, it's a pain. We can't do much about it (except maybe adding the date to the filenames). But a link to the latest snapshot is on very first page: https://bintray.com/aranym/aranym-files/snapshots/_latestVersion#files

**Comment by vinriviere** [comment]
I've just tested Pupul with latest ARAnyM snapshot: It works 😃
Great job, ARAnyM team.
Finally what has been fixed in ARAnyM?

**Comment by th-otto** [comment]
As for now, only the value that is used for silence, when there is not enough audio data available. But good to know that it works now, so this issue is maybe solved for now.

I still want to investigate the calculation of the buffptr address while audio data is playing, having only return values that match half of the buffer size seems rather suspicious. Didn't someone want to supply a test program for this purpose?

**Comment by vinriviere** [comment]
Yesterday, after testing the fixed ARAnyM, I recorded the ARAnyM audio output with Audacity. It was perfectly continuous, without any glitch. It's rather strange that such "silence value" is not visible in the wave output. Even if it works fine now, that's still a mystery.
AFAIK, no one provided a simpler testcase.

**Comment by th-otto** [comment]
i'm not entirely sure how audacity catches the audio output, but it is possible that merely the recording has an influence of such delays. Typically (on linux at least) audio output is sent to a pulseaudio server, and that may cause some latency. When i implemented the fix, i temporarily added some debug printf, and found that (while running pupul) the function was called for ~500 bytes of silence on linux, and ~2000 bytes on windows. The shorter delay on linux was hardly noticable, even without the fix.
